compiler/rustc_const_eval/src/const_eval/valtrees.rs

```diff
@@ -0,0 +1,361 @@
+                    let len = valtree.unwrap_branch().len() as u64;
+                    create_mplace_from_layout(ecx, *inner_ty, Some(len))
+                }
+                _ => create_mplace_from_layout(ecx, *inner_ty, None),
+            };
+            debug!(?pointee_place);
+
+            fill_place_recursively(ecx, &mut pointee_place, valtree);
+            dump_place(ecx, pointee_place.into());
+            intern_const_alloc_recursive(ecx, InternKind::Constant, &pointee_place).unwrap();
+
+            let imm = match inner_ty.kind() {
+                ty::Slice(_) | ty::Str => {
+                    let len = valtree.unwrap_branch().len();
+                    let len_scalar = ScalarMaybeUninit::Scalar(Scalar::from_u64(len as u64));
+
+                    Immediate::ScalarPair(
+                        ScalarMaybeUninit::from_maybe_pointer((*pointee_place).ptr, &tcx),
+                        len_scalar,
+                    )
+                }
+                _ => pointee_place.to_ref(&tcx),
+            };
+            debug!(?imm);
+
+            ecx.write_immediate(imm, &(*place).into()).unwrap();
+        }
+        ty::Adt(_, _) | ty::Tuple(_) | ty::Array(_, _) | ty::Str => {
+            let branches = valtree.unwrap_branch();
+
+            // Need to downcast place for enums
+            let (place_adjusted, branches, variant_idx) = match ty.kind() {
+                ty::Adt(def, _) if def.is_enum() => {
+                    // First element of valtree corresponds to variant
+                    let scalar_int = branches[0].unwrap_leaf();
+                    let variant_idx = VariantIdx::from_u32(scalar_int.try_to_u32().unwrap());
+                    let variant = def.variant(variant_idx);
+                    debug!(?variant);
+
+                    (
+                        place.project_downcast(ecx, variant_idx).unwrap(),
+                        &branches[1..],
+                        Some(variant_idx),
+                    )
+                }
+                _ => (*place, branches, None),
+            };
+            debug!(?place_adjusted, ?branches);
+
+            // Create the places for the fields and fill them recursively
+            for (i, inner_valtree) in branches.iter().enumerate() {
+                debug!(?i, ?inner_valtree);
+
+                let mut place_inner = match *ty.kind() {
+                    ty::Adt(def, substs) if !def.is_enum() => {
+                        let field = &def.variant(VariantIdx::from_usize(0)).fields[i];
+                        let field_ty = field.ty(tcx, substs);
+                        let projection_elem = ProjectionElem::Field(Field::from_usize(i), field_ty);
+
+                        ecx.mplace_projection(&place_adjusted, projection_elem).unwrap()
+                    }
+                    ty::Adt(_, _) | ty::Tuple(_) => ecx.mplace_field(&place_adjusted, i).unwrap(),
+                    ty::Array(_, _) | ty::Str => {
```

Comment:
This still fails. I'm currently getting the following when compiling std during the conversion for a valtree with underlying type `&[&str; 1]`. In the recursive call for `str` we get:

```
│ │ │ │ │ ├─┐rustc_const_eval::const_eval::valtrees::fill_place_recursively place=MPlaceTy { mplace: MemPlace { ptr: alloc1411, align: Align { pow2: 0 }, meta: Meta(0x000000000000002a) }, layout: TyAndLayout { ty: str, layout: Layout { fields: Array { stride: Size { raw: 1 }, count: 0 }, variants: Single { index: 0 }, abi: Aggregate { sized: false }, largest_niche: None, align: AbiAndPrefAlign { abi: Align { pow2: 0 }, pref: Align { pow2: 0 } }, size: Size { raw: 0 } } } }
``` 

(Had to manually add the meta data for the pointer in `create_mplace_from_layout`, which seems fishy, but I can't call deref_operand since the memory is uninit at this stage).

As a comparison in the valtree creation we have this call to `slice_branches` (which works):

```
│ │ │ │ │ │ │ │ ├─┐rustc_const_eval::const_eval::valtrees::slice_branches place=MPlaceTy { mplace: MemPlace { ptr: alloc1408, align: Align { pow2: 0 }, meta: Meta(0x000000000000002a) }, layout: TyAndLayout { ty: str, layout: Layout { fields: Array { stride: Size { raw: 1 }, count: 0 }, variants: Single { index: 0 }, abi: Aggregate { sized: false }, largest_niche: None, align: AbiAndPrefAlign { abi: Align { pow2: 0 }, pref: Align { pow2: 0 } }, size: Size { raw: 0 } } } }
```

but in the bottom call:

```
│ │ │ │ │ │ ├─┐rustc_const_eval::const_eval::valtrees::fill_place_recursively place=MPlaceTy { mplace: MemPlace { ptr: alloc1411, align: Align { pow2: 0 }, meta: None }, layout: TyAndLayout { ty: u8, layout: Layout { fields: Primitive, variants: Single { index: 0 }, abi: Scalar(Initialized { value: Int(I8, false), valid_range: 0..=255 }), largest_niche: None, align: AbiAndPrefAlign { abi: Align { pow2: 0 }, pref: Align { pow2: 0 } }, size: Size { raw: 1 } } } }, valtree=Leaf(0x66)
```

the write_immediate call fails with the error:

```
│ │ │ │ │ │ │ │ │ ├─┐rustc_const_eval::interpret::place::write_immediate_to_mplace_no_validate value=Scalar(0x66), dest=MPlaceTy { mplace: MemPlace { ptr: alloc1411, align: Align { pow2: 0 }, meta: None }, layout: TyAndLayout { ty: u8, layout: Layout { fields: Primitive, variants: Single { index: 0 }, abi: Scalar(Initialized { value: Int(I8, false), valid_range: 0..=255 }), largest_niche: None, align: AbiAndPrefAlign { abi: Align { pow2: 0 }, pref: Align { pow2: 0 } }, size: Size { raw: 1 } } }
```

```
InterpErrorInfo(InterpErrorInfoInner { kind: memory access failed: alloc1411 has size 0, so pointer to 1 byte starting at offset 0 is out-of-bounds, backtrace: None })
```

Trying to figure out why this isn't working, but if one of you know what's going on, I'd appreciate your thoughts.